libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt

```diff
@@ -214,38 +221,51 @@ internal class MapboxRerouteController @VisibleForTesting constructor(
         callback: NavigationRerouteController.RoutesCallback,
         routeOptions: RouteOptions
     ) {
-        requestId = directionsSession.requestRoutes(
-            routeOptions,
-            object : NavigationRouterCallback {
-                override fun onRoutesReady(
-                    routes: List<NavigationRoute>,
-                    routerOrigin: RouterOrigin
-                ) {
-                    mainJobController.scope.launch {
-                        state = RerouteState.RouteFetched(routerOrigin)
-                        state = RerouteState.Idle
-                        callback.onNewRoutes(routes, routerOrigin)
-                    }
-                }
+        rerouteScope.launch {
+            val result = suspendCancellableCoroutine<Pair<List<NavigationRoute>, RouterOrigin>?> { cont ->
```

Comment:
```suggestion
            val result = suspendCoroutine<Pair<List<NavigationRoute>, RouterOrigin>?> { cont ->
```
This coroutine cannot be cancelable because we rely on the callbacks to return so that we can set the correct state. Right now it doesn't truly matter if we use `suspendCoroutine` or `suspendCancellableCoroutine` because there are no suspensions that could be canceled, but just for clarity or in case the implementation changes, we should use a non-cancelable coroutine.

Either that, or we should add a
```
cont.invokeOnCancellation { }
```
block where we set the correct state.

Reply:
@LukasPaczos even if you use `suspendCoroutine`, callbacks will not be invoked if `rerouteScope` was canceled for any reason.

Reply:
That makes sense, thanks @Zayankovsky. In this case looks like we'll need to use `invokeOnCancellation` block.

Reply:
Wait, this isn't truly what we're looking for @Zayankovsky. We don't need the coroutine to proceed and consume the result, we just need to make sure that callbacks of `NavigationRouterCallback` fire so that we can schedule state updates. Since there are no suspensions in that block https://github.com/mapbox/mapbox-navigation-android/pull/6693#discussion_r1042238973 still holds true, doesn't it?

Reply:
Sounds good :+1:

Reply:
If you use `invokeOnCancellation` then you should move state updates from `NavigationRouterCallback` callbacks to after `suspendCancellableCoroutine`. Otherwise there may be extra updates (from one of the callback and from `invokeOnCancellation`).

Reply:
You are right. Thanks.